src/fsharp/FSharp.Core/control.fs

```diff
@@ -2084,9 +2084,8 @@ namespace Microsoft.FSharp.Control
                             // If it isn't, then create it, and go back to the start to 
                             // check arrivals again.
                             match pulse with
-                            | null -> 
-                                if timeout >= 0 || cancellationSupported then 
-                                    ensurePulse() |> ignore
+                            | null when timeout >= 0 || cancellationSupported ->
+                                ensurePulse() |> ignore
                                 return! processFirstArrival()
```

Comment:
I think the problem is that we need to call `ensurePulse` and than check arrivals a single time again before doing the actual wait (on the value returned by `ensurePulse`). While I think this patch is doing just that, it makes the code even more unreadable. But that might be some refactoring for another time.